File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        solve(board);
    }
    bool solve(vector<vector<char>>& board){
        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
                if(board[i][j]=='.'){
                    for(char c='1';c<='9';c++){
                        if(isValid(board,i,j,c)){
                            board[i][j]=c;
                            if(solve(board)) return true;
                            else board[i][j]='.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    bool isValid(vector<vector<char>>& board,int i,int j,char c){
        for(int a=0;a<9;a++){
            if(board[a][j]==c) return false;
            if(board[i][a]==c) return false;
            if(board[3*(i/3)+a/3][3*(j/3)+(a%3)]==c) return false;   
        }
        return true;
    }
};

File: 0037-sudoku-solver/0037-sudoku-solver.java
class Solution {

    private boolean isSafe(char[][] board, int row, int col, char digit) {

        // Check row
        for (int j = 0; j < 9; j++) {
            if (board[row][j] == digit)
                return false;
        }

        // Check column
        for (int i = 0; i < 9; i++) {
            if (board[i][col] == digit)
                return false;
        }

        // Check 3x3 box
        int startRow = (row / 3) * 3;
        int startCol = (col / 3) * 3;

        for (int i = startRow; i < startRow + 3; i++) {
            for (int j = startCol; j < startCol + 3; j++) {
                if (board[i][j] == digit)
                    return false;
            }
        }

        return true;
    }

    private boolean solve(char[][] board) {

        // Find the first empty cell
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {

                if (board[i][j] == '.') {

                    // Try digits 1 to 9
                    for (char digit = '1'; digit <= '9'; digit++) {

                        if (isSafe(board, i, j, digit)) {

                            // Choose
                            board[i][j] = digit;

                            // Explore
                            if (solve(board))
                                return true;

                            // Backtrack
                            board[i][j] = '.';
                        }
                    }

                    // No digit works
                    return false;
                }
            }
        }

        // No empty cell left
        return true;
    }

    public void solveSudoku(char[][] board) {
        solve(board);
    }
}